Dispatch backend-specific TensorOptions-based 'factory' functions via…

gchanan opened the pull request:
… Type.

This allows one to write a cpu/cuda split 'factory' function that uses TensorOptions.
Also move all remaining native_functions with either function or method variants that use Type to use TensorOptions.
Thus, there are no more Types in the public function / method API.

I believe there is a _lot_ of opportunity for cleanup here, as the old tensor, th_tensor, native_tensor and sparse variants can probably be removed, but let's do that in a follow-on patch.

ezyang commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -33,8 +33,8 @@ SparseTensorImpl::SparseTensorImpl(at::TensorTypeId type_id, const caffe2::TypeM
     , denseDims_(0)
-    , indices_(globalContext().getNonVariableTypeOpt(sparseTensorIdToDenseBackend(type_id), ScalarType::Long)->tensor({1, 0}))
-    , values_(globalContext().getNonVariableTypeOpt(sparseTensorIdToDenseBackend(type_id), dataTypeToScalarType(data_type.id()))->tensor()) {}
+    , indices_(at::empty({1, 0}, TensorOptions(false).device(sparseTensorIdToDeviceType(type_id)).dtype(ScalarType::Long)))
```
Aaawkard! :)

gchanan replied:
It's better than before IMO :).

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -832,7 +835,7 @@ def process_option(option, output_options):
         # "s_" for "same size".
         option['method_prefix_derived'] = '' if broadcast_arg is None else 's_'
-        option['device_guard_declaration'] = device_guard(option, formals)
+        option['device_guard_declaration'] = device_guard(option, formals, False, False)
```
boooolean blindness

gchanan replied:
idk why we were defaulting arguments with a single call site.

ezyang commented on aten/src/ATen/native/cudnn/RNN.cpp:
```diff
@@ -1113,21 +1113,21 @@ struct DropoutState {
     int64_t seed = at::empty({}, at::kLong).random_().item<int64_t>();
     state.buffer = at::_cudnn_init_dropout_state(
-      tp.toScalarType(at::kByte), dropout_p, train, seed);
+      dropout_p, train, seed, options.dtype(at::kByte));
```
Eww, mutation.

gchanan replied:
well, the param is a value type so it ends up not mattering, but you are correct that this is a subtlety of the `TensorOptions` API.

ezyang commented on tools/autograd/gen_variable_factories.py:
```diff
@@ -34,7 +34,8 @@ def gen_variable_factories(out, declarations, template_path):
     for decl in declarations:
         has_tensor_options = any(a["simple_type"] == "TensorOptions" for a in decl["arguments"])
-        if has_tensor_options or decl["name"].endswith("_like"):
+        is_namespace_fn = 'namespace' in decl['method_of']
```
Oh, that's interesting. What does `method_of` do again?

gchanan replied:
This is saying we should only generate a variable factory corresponding to a tensor factory function if it is actually a namespace function, i.e. has variant: function.  The previous code would generate a variable factory if it had a TensorOptions parameter, but this had a bad effect when combined with the `variants: []` concept when I moved some of the Type-parameter functions to be TensorOptions-parameter functions (i.e. it would generate variable factories even though a corresponding factory function didn't exist).

gchanan commented:
problem appears to be that Type method aren't getting generated in VariableType.

gchanan committed:
Dispatch backend-specific TensorOptions-based 'factory' functions via… (#12071)

Summary:
… Type.

This allows one to write a cpu/cuda split 'factory' function that uses TensorOptions.
Also move all remaining native_functions with either function or method variants that use Type to use TensorOptions.
Thus, there are no more Types in the public function / method API.

I believe there is a _lot_ of opportunity for cleanup here, as the old tensor, th_tensor, native_tensor and sparse variants can probably be removed, but let's do that in a follow-on patch.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12071

Reviewed By: ezyang

Differential Revision: D10041600

fbshipit-source-id: 637e2075f7e72a98332a4827ae45adcd8c400078
